Provide better error message for tracepoints with undefined struct types

I think we can somehow provide better information for users when there are missing includes in tracepoint usage. For example

```
# bpftrace -e 'tracepoint:syscalls:sys_enter_io_setup {printf("%u %x %p\n", pid, args->nr_events, args->ctxp)}'
definitions.h:10:3: error: unknown type name 'aio_context_t'

```

The program above needs `uapi/linux/aio_abi.h` to work, but the error message don't give anything actionable to the user (not even a tip that maybe they should include a Kernel header).

_Originated from https://github.com/iovisor/bpftrace/issues/604#issuecomment-499699673_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,7 +26,7 @@ and this project adheres to
   - [#1476](https://github.com/iovisor/bpftrace/pull/1476)
 - List retprobes
   - [#1484](https://github.com/iovisor/bpftrace/pull/1484)
-- Resolve unknown typedefs using BTF
+- Resolve unknown typedefs using BTF and give a hint when a type cannot be found
   - [#1485](https://github.com/iovisor/bpftrace/pull/1485)
 
 #### Changed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,8 @@ and this project adheres to
   - [#1476](https://github.com/iovisor/bpftrace/pull/1476)
 - List retprobes
   - [#1484](https://github.com/iovisor/bpftrace/pull/1484)
+- Resolve unknown typedefs using BTF
+  - [#1485](https://github.com/iovisor/bpftrace/pull/1485)
 
 #### Changed
 - Warn if using `print` on `stats` maps with top and div arguments

--- a/src/clang_parser.cpp
+++ b/src/clang_parser.cpp
@@ -386,13 +386,6 @@ bool ClangParser::ClangParserHandler::check_diagnostics(
   return true;
 }
 
-bool ClangParser::ClangParserHandler::check_diagnostics(
-    const std::string &input)
-{
-  std::vector<std::string> msgs;
-  return check_diagnostics(input, msgs, true);
-}
-
 CXCursor ClangParser::ClangParserHandler::get_translation_unit_cursor() {
   return clang_getTranslationUnitCursor(translation_unit);
 }
@@ -529,15 +522,11 @@ std::unordered_set<std::string> ClangParser::get_incomplete_types(
   //   unknown type name 'type_t'
   // that imply an unresolved typedef of type_t. This cannot be done below in
   // clang_visitChildren since clang does not have the unknown type names.
-  const std::string unknown_type_msg = "unknown type name \'";
-  for (auto &msg : diag_msgs)
+  for (const auto &msg : diag_msgs)
   {
-    if (msg.find(unknown_type_msg) == 0)
-    {
-      type_data.incomplete_types.emplace(
-          msg.substr(unknown_type_msg.length(),
-                     msg.length() - unknown_type_msg.length() - 1));
-    }
+    auto unknown_type = get_unknown_type(msg);
+    if (unknown_type)
+      type_data.incomplete_types.emplace(unknown_type.value());
   }
 
   CXCursor cursor = handler.get_translation_unit_cursor();
@@ -667,11 +656,40 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     return false;
   }
 
-  if (!handler.check_diagnostics(input))
+  std::vector<std::string> error_msgs;
+  if (!handler.check_diagnostics(input, error_msgs, true))
+  {
+    for (auto &msg : error_msgs)
+    {
+      if (get_unknown_type(msg) != "" && !bpftrace.force_btf_)
+      {
+        LOG(ERROR) << "Try running with --btf to force BTF processing or "
+                      "include headers with missing type definitions";
+      }
+    }
     return false;
+  }
 
   CXCursor cursor = handler.get_translation_unit_cursor();
   return visit_children(cursor, bpftrace);
+}
+
+/*
+ * Parse the given Clang diagnostics message and if it has the form:
+ *   unknown type name 'type_t'
+ * return type_t.
+ */
+std::optional<std::string> ClangParser::ClangParser::get_unknown_type(
+    const std::string &diagnostic_msg)
+{
+  const std::string unknown_type_msg = "unknown type name \'";
+  if (diagnostic_msg.find(unknown_type_msg) == 0)
+  {
+    return diagnostic_msg.substr(unknown_type_msg.length(),
+                                 diagnostic_msg.length() -
+                                     unknown_type_msg.length() - 1);
+  }
+  return {};
 }
 
 } // namespace bpftrace

--- a/src/clang_parser.cpp
+++ b/src/clang_parser.cpp
@@ -363,20 +363,34 @@ CXErrorCode ClangParser::ClangParserHandler::parse_translation_unit(
 
 bool ClangParser::ClangParserHandler::check_diagnostics(
     const std::string &input,
+    std::vector<std::string> &error_msgs,
     bool bail_on_error)
 {
   for (unsigned int i=0; i < clang_getNumDiagnostics(get_translation_unit()); i++) {
     CXDiagnostic diag = clang_getDiagnostic(get_translation_unit(), i);
     CXDiagnosticSeverity severity = clang_getDiagnosticSeverity(diag);
+    std::string msg = clang_getCString(clang_getDiagnosticSpelling(diag));
+    error_msgs.push_back(msg);
+
     if ((bail_on_error && severity == CXDiagnostic_Error) ||
         severity == CXDiagnostic_Fatal)
     {
+      // Do not fail on "too many errors"
+      if (!bail_on_error && msg == "too many errors emitted, stopping now")
+        return true;
       if (bt_debug >= DebugLevel::kDebug)
         LOG(ERROR) << "Input (" << input.size() << "): " << input;
       return false;
     }
   }
   return true;
+}
+
+bool ClangParser::ClangParserHandler::check_diagnostics(
+    const std::string &input)
+{
+  std::vector<std::string> msgs;
+  return check_diagnostics(input, msgs, true);
 }
 
 CXCursor ClangParser::ClangParserHandler::get_translation_unit_cursor() {
@@ -461,7 +475,8 @@ bool ClangParser::visit_children(CXCursor &cursor, BPFtrace &bpftrace)
 std::unordered_set<std::string> ClangParser::get_incomplete_types(
     const std::string &input,
     std::vector<CXUnsavedFile> &unsaved_files,
-    const std::vector<const char *> &args)
+    const std::vector<const char *> &args,
+    const std::unordered_set<std::string> &complete_types)
 {
   if (input.empty())
     return {};
@@ -495,9 +510,11 @@ std::unordered_set<std::string> ClangParser::get_incomplete_types(
     return {};
   }
 
-  // Don't bail on errors (ie incomplete structs) because our goal
-  // is to enumerate all such errors
-  if (!handler.check_diagnostics(input, /* bail_on_error= */ false))
+  // Don't bail on errors (ie incomplete structs) because our goal is to
+  // enumerate all such errors. Instead, collect error messages for later
+  // analysis.
+  std::vector<std::string> diag_msgs;
+  if (!handler.check_diagnostics(input, diag_msgs, false))
     return {};
 
   struct TypeData
@@ -505,6 +522,23 @@ std::unordered_set<std::string> ClangParser::get_incomplete_types(
     std::unordered_set<std::string> complete_types;
     std::unordered_set<std::string> incomplete_types;
   } type_data;
+  // Initialize to already defined types
+  type_data.complete_types = complete_types;
+
+  // Search for error messages of the form:
+  //   unknown type name 'type_t'
+  // that imply an unresolved typedef of type_t. This cannot be done below in
+  // clang_visitChildren since clang does not have the unknown type names.
+  const std::string unknown_type_msg = "unknown type name \'";
+  for (auto &msg : diag_msgs)
+  {
+    if (msg.find(unknown_type_msg) == 0)
+    {
+      type_data.incomplete_types.emplace(
+          msg.substr(unknown_type_msg.length(),
+                     msg.length() - unknown_type_msg.length() - 1));
+    }
+  }
 
   CXCursor cursor = handler.get_translation_unit_cursor();
   clang_visitChildren(
@@ -550,7 +584,7 @@ std::unordered_set<std::string> ClangParser::get_incomplete_types(
 
 bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<std::string> extra_flags)
 {
-  auto input = program->c_definitions;
+  auto input = "#include <__btf_generated_header.h>\n" + program->c_definitions;
 
   auto input_files = getTranslationUnitFiles(CXUnsavedFile{
       .Filename = "definitions.h",
@@ -569,7 +603,7 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     args.push_back(flag.c_str());
   }
 
-  bool process_btf = input.empty() ||
+  bool process_btf = program->c_definitions.empty() ||
                      (bpftrace.force_btf_ && bpftrace.btf_.has_data());
 
   // We set these args early because some systems may not have <linux/types.h>
@@ -584,45 +618,45 @@ bool ClangParser::parse(ast::Program *program, BPFtrace &bpftrace, std::vector<s
     args.push_back("-D__CLANG_WORKAROUNDS_H");
   }
 
-  auto incomplete_types = get_incomplete_types(input, input_files, args);
-  bpftrace.btf_set_.insert(incomplete_types.cbegin(), incomplete_types.cend());
+  // The generated BTF header is initially empty
+  std::string btf_cdef;
+  input_files.emplace_back(CXUnsavedFile{
+      .Filename = "/bpftrace/include/__btf_generated_header.h",
+      .Contents = btf_cdef.c_str(),
+      .Length = btf_cdef.size(),
+  });
 
   ClangParserHandler handler;
+  bool check_additional_types = true;
+  while (check_additional_types && process_btf)
+  {
+    auto incomplete_types = get_incomplete_types(
+        input, input_files, args, bpftrace.btf_set_);
+    size_t types_cnt = bpftrace.btf_set_.size();
+    bpftrace.btf_set_.insert(incomplete_types.cbegin(),
+                             incomplete_types.cend());
+
+    // Update generated header with current BTF types
+    btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
+    input_files.back() = CXUnsavedFile{
+      .Filename = "/bpftrace/include/__btf_generated_header.h",
+      .Contents = btf_cdef.c_str(),
+      .Length = btf_cdef.size(),
+    };
+
+    // If additional BTF types were found, we need to repeat the process since
+    // that might have introduced some new unresolved typedefs.
+    check_additional_types = types_cnt != bpftrace.btf_set_.size();
+  }
+
   CXErrorCode error;
-
-  if (process_btf)
-  {
-    auto btf_and_input = "#include <__btf_generated_header.h>\n" + input;
-    auto btf_and_input_files = getTranslationUnitFiles(
-        CXUnsavedFile{ .Filename = "definitions.h",
-                       .Contents = btf_and_input.c_str(),
-                       .Length = btf_and_input.size() });
-
-    auto btf_cdef = bpftrace.btf_.c_def(bpftrace.btf_set_);
-    btf_and_input_files.emplace_back(CXUnsavedFile{
-        .Filename = "/bpftrace/include/__btf_generated_header.h",
-        .Contents = btf_cdef.c_str(),
-        .Length = btf_cdef.size(),
-    });
-
-    error = handler.parse_translation_unit(
-        "definitions.h",
-        args.data(),
-        args.size(),
-        btf_and_input_files.data(),
-        btf_and_input_files.size(),
-        CXTranslationUnit_DetailedPreprocessingRecord);
-  }
-  else
-  {
-    error = handler.parse_translation_unit(
-        "definitions.h",
-        args.data(),
-        args.size(),
-        input_files.data(),
-        input_files.size(),
-        CXTranslationUnit_DetailedPreprocessingRecord);
-  }
+  error = handler.parse_translation_unit(
+      "definitions.h",
+      args.data(),
+      args.size(),
+      input_files.data(),
+      input_files.size(),
+      CXTranslationUnit_DetailedPreprocessingRecord);
 
   if (error)
   {

--- a/src/clang_parser.h
+++ b/src/clang_parser.h
@@ -41,6 +41,9 @@ private:
       const std::vector<const char *> &args,
       const std::unordered_set<std::string> &complete_types);
 
+  static std::optional<std::string> get_unknown_type(
+      const std::string &diagnostic_msg);
+
   class ClangParserHandler
   {
   public:
@@ -65,10 +68,6 @@ private:
     bool check_diagnostics(const std::string &input,
                            std::vector<std::string> &error_msgs,
                            bool bail_on_error);
-    /*
-     * Runs the above method with bail_on_error=false and throws away messages.
-     */
-    bool check_diagnostics(const std::string &input);
 
     CXCursor get_translation_unit_cursor();
 

--- a/src/clang_parser.h
+++ b/src/clang_parser.h
@@ -24,8 +24,11 @@ private:
   bool visit_children(CXCursor &cursor, BPFtrace &bpftrace);
   /*
    * The user might have written some struct definitions that rely on types
-   * supplied by BTF data. This method will pull out any forward-declared /
-   * incomplete struct definitions and return the types (in string form) of
+   * supplied by BTF data. Also, some types may be defined by typedefs that are
+   * in non-included headers, which causes problems to clang.
+   *
+   * This method will pull out any forward-declared / incomplete struct
+   * and typedef definitions and return the types (in string form) of
    * the unresolved types.
    *
    * Note that this method does not report "errors". This is because the user
@@ -35,7 +38,8 @@ private:
   std::unordered_set<std::string> get_incomplete_types(
       const std::string &input,
       std::vector<CXUnsavedFile> &unsaved_files,
-      const std::vector<const char *> &args);
+      const std::vector<const char *> &args,
+      const std::unordered_set<std::string> &complete_types);
 
   class ClangParserHandler
   {
@@ -53,7 +57,18 @@ private:
                                        unsigned num_unsaved_files,
                                        unsigned options);
 
-    bool check_diagnostics(const std::string &input, bool bail_on_error = true);
+    /*
+     * Check diagnostics and collect all error messages.
+     * Return true if an error occurred. If bail_on_error is false, only fail
+     * on fatal errors.
+     */
+    bool check_diagnostics(const std::string &input,
+                           std::vector<std::string> &error_msgs,
+                           bool bail_on_error);
+    /*
+     * Runs the above method with bail_on_error=false and throws away messages.
+     */
+    bool check_diagnostics(const std::string &input);
 
     CXCursor get_translation_unit_cursor();
 

--- a/tests/clang_parser.cpp
+++ b/tests/clang_parser.cpp
@@ -221,8 +221,8 @@ TEST(clang_parser, nested_struct_no_type)
   // since they are called bar and baz
   parse("struct Foo { struct { int x; } bar; union { int y; } baz; }", bpftrace);
 
-  std::string bar = "struct Foo::(anonymous at definitions.h:1:14)";
-  std::string baz = "union Foo::(anonymous at definitions.h:1:37)";
+  std::string bar = "struct Foo::(anonymous at definitions.h:2:14)";
+  std::string baz = "union Foo::(anonymous at definitions.h:2:37)";
   StructMap &structs = bpftrace.structs_;
 
   ASSERT_EQ(structs.size(), 3U);
@@ -654,6 +654,27 @@ TEST_F(clang_parser_btf, btf_variable_field_struct)
   // struct Foo2 should be added by @x1->foo2
   EXPECT_NE(bpftrace.btf_set_.find("struct Foo2"), bpftrace.btf_set_.end());
   EXPECT_NE(bpftrace.btf_set_.find("struct Foo3"), bpftrace.btf_set_.end());
+}
+
+TEST(clang_parser, btf_unresolved_typedef)
+{
+  // size_t is defined in stddef.h, but if we have BTF, it should be possible to
+  // extract it from there
+  BPFtrace bpftrace;
+  bpftrace.force_btf_ = true;
+  parse("struct Foo { size_t x; };", bpftrace);
+
+  StructMap &structs = bpftrace.structs_;
+
+  ASSERT_EQ(structs.count("struct Foo"), 1U);
+
+  EXPECT_EQ(structs["struct Foo"].size, 8);
+  ASSERT_EQ(structs["struct Foo"].fields.size(), 1U);
+  ASSERT_EQ(structs["struct Foo"].fields.count("x"), 1U);
+
+  EXPECT_EQ(structs["struct Foo"].fields["x"].type.type, Type::integer);
+  EXPECT_EQ(structs["struct Foo"].fields["x"].type.size, 8U);
+  EXPECT_EQ(structs["struct Foo"].fields["x"].offset, 0);
 }
 #endif // HAVE_LIBBPF_BTF_DUMP
 

--- a/tests/codegen/llvm/struct_nested_struct_anon_1.ll
+++ b/tests/codegen/llvm/struct_nested_struct_anon_1.ll
@@ -10,7 +10,7 @@ define i64 @"kprobe:f"(i8*) section "s_kprobe:f_1" {
 entry:
   %"@x_val" = alloca i64
   %"@x_key" = alloca i64
-  %"struct Foo::(anonymous at definitions.h:1:14).x" = alloca i32
+  %"struct Foo::(anonymous at definitions.h:2:14).x" = alloca i32
   %"$foo" = alloca [4 x i8]
   %1 = bitcast [4 x i8]* %"$foo" to i8*
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* %1)
@@ -23,12 +23,12 @@ entry:
   call void @llvm.memcpy.p0i8.p64i8.i64(i8* align 1 %4, i8 addrspace(64)* align 1 %5, i64 4, i1 false)
   %6 = add [4 x i8]* %"$foo", i64 0
   %7 = add [4 x i8]* %6, i64 0
-  %8 = bitcast i32* %"struct Foo::(anonymous at definitions.h:1:14).x" to i8*
+  %8 = bitcast i32* %"struct Foo::(anonymous at definitions.h:2:14).x" to i8*
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* %8)
-  %probe_read = call i64 inttoptr (i64 4 to i64 (i32*, i32, [4 x i8]*)*)(i32* %"struct Foo::(anonymous at definitions.h:1:14).x", i32 4, [4 x i8]* %7)
-  %9 = load i32, i32* %"struct Foo::(anonymous at definitions.h:1:14).x"
+  %probe_read = call i64 inttoptr (i64 4 to i64 (i32*, i32, [4 x i8]*)*)(i32* %"struct Foo::(anonymous at definitions.h:2:14).x", i32 4, [4 x i8]* %7)
+  %9 = load i32, i32* %"struct Foo::(anonymous at definitions.h:2:14).x"
   %10 = sext i32 %9 to i64
-  %11 = bitcast i32* %"struct Foo::(anonymous at definitions.h:1:14).x" to i8*
+  %11 = bitcast i32* %"struct Foo::(anonymous at definitions.h:2:14).x" to i8*
   call void @llvm.lifetime.end.p0i8(i64 -1, i8* %11)
   %12 = bitcast i64* %"@x_key" to i8*
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* %12)

--- a/tests/codegen/llvm/struct_nested_struct_anon_2.ll
+++ b/tests/codegen/llvm/struct_nested_struct_anon_2.ll
@@ -10,7 +10,7 @@ define i64 @"kprobe:f"(i8*) section "s_kprobe:f_1" {
 entry:
   %"@x_val" = alloca i64
   %"@x_key" = alloca i64
-  %"struct Foo::(anonymous at definitions.h:1:14).x" = alloca i32
+  %"struct Foo::(anonymous at definitions.h:2:14).x" = alloca i32
   %"$foo" = alloca i64
   %1 = bitcast i64* %"$foo" to i8*
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* %1)
@@ -21,12 +21,12 @@ entry:
   %3 = load i64, i64* %"$foo"
   %4 = add i64 %3, 0
   %5 = add i64 %4, 0
-  %6 = bitcast i32* %"struct Foo::(anonymous at definitions.h:1:14).x" to i8*
+  %6 = bitcast i32* %"struct Foo::(anonymous at definitions.h:2:14).x" to i8*
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* %6)
-  %probe_read = call i64 inttoptr (i64 4 to i64 (i32*, i32, i64)*)(i32* %"struct Foo::(anonymous at definitions.h:1:14).x", i32 4, i64 %5)
-  %7 = load i32, i32* %"struct Foo::(anonymous at definitions.h:1:14).x"
+  %probe_read = call i64 inttoptr (i64 4 to i64 (i32*, i32, i64)*)(i32* %"struct Foo::(anonymous at definitions.h:2:14).x", i32 4, i64 %5)
+  %7 = load i32, i32* %"struct Foo::(anonymous at definitions.h:2:14).x"
   %8 = sext i32 %7 to i64
-  %9 = bitcast i32* %"struct Foo::(anonymous at definitions.h:1:14).x" to i8*
+  %9 = bitcast i32* %"struct Foo::(anonymous at definitions.h:2:14).x" to i8*
   call void @llvm.lifetime.end.p0i8(i64 -1, i8* %9)
   %10 = bitcast i64* %"@x_key" to i8*
   call void @llvm.lifetime.start.p0i8(i64 -1, i8* %10)

